fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+ *
+ * This function performs a key derivation operation using the given
+ * private key and peer public key. The resulting shared secret is
+ * allocated dynamically and must be freed by the caller.
+ *
+ * @param[in] key The private key used for derivation.
+ * @param[in] peer The peer's public key.
+ * @param[out] shared Pointer to the derived shared secret (allocated).
+ * @param[out] shared_len Length of the derived shared secret.
+ *
+ * @note The caller is responsible for freeing the memory allocated
+ *       for `shared` using `OPENSSL_free()`.
+ */
+static void do_derive(EVP_PKEY *key, EVP_PKEY *peer, uint8_t **shared, size_t *shared_len)
+{
+    EVP_PKEY_CTX *ctx = NULL;
+
+    *shared = NULL;
+    *shared_len = 0;
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);
+    if (ctx == NULL) {
+        fprintf(stderr, "failed to create keygen context\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_derive_init(ctx)) {
+        fprintf(stderr, "failed to init derive context\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_derive_set_peer(ctx, peer)) {
+        fprintf(stderr, "failed to set peer\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_derive(ctx, NULL, shared_len)) {
+        fprintf(stderr, "Derive failed 1\n");
+        goto err;
+    }
+
+    if (*shared_len == 0)
+        goto err;
+
+    *shared = OPENSSL_zalloc(*shared_len);
+    if (*shared == NULL) {
+        fprintf(stderr, "Failed to alloc\n");
+        goto err;
+    }
+    if (!EVP_PKEY_derive(ctx, *shared, shared_len)) {
+        fprintf(stderr, "Derive failed 2\n");
+        OPENSSL_free(*shared);
+        *shared = NULL;
+        *shared_len = 0;
+        goto err;
+    }
+err:
+    EVP_PKEY_CTX_free(ctx);
+}
+
+/**
+ * @brief Performs a key exchange using ML-KEM.
+ *
+ * This function derives shared secrets using the provided key pairs.
+ * It calls `do_derive()` to compute shared secrets for both participants
+ * and frees the allocated memory for the shared secrets.
+ *
+ * @param[out] buf Unused output buffer (reserved for future use).
+ * @param[out] len Unused output length (reserved for future use).
+ * @param[in] key1 First key (typically Alice's key).
+ * @param[in] key2 Second key (typically Bob's key).
+ * @param[out] out1 Unused output parameter (reserved for future use).
+ * @param[out] out2 Unused output parameter (reserved for future use).
+ *
+ * @note Currently, this function does not validate whether the derived
+ *       shared secrets match. A check should be added when ML-KEM
+ *       supports this.
+ */
+static void mlkem_kex(uint8_t **buf, size_t *len, void *key1, void *key2,
+                      void **out1, void **out2)
+{
+    EVP_PKEY *alice = (EVP_PKEY *)key1;
+    EVP_PKEY *bob = (EVP_PKEY *)key2;
+    size_t boblen, alicelen;
+    uint8_t *bobshare = NULL;
+    uint8_t *aliceshare = NULL;
+
+    do_derive(alice, bob, &aliceshare, &alicelen);
+    do_derive(bob, alice, &bobshare, &boblen);
```

Comment:
So this should also always fail. With KEMs, Alice sends Bob her public key, Bob encapsulates, and then Alice decapsulates.

Reply:
correct